### ISSUE TITLE
CMP PKIHeader.protectionAlg incorrect for RSA

### ISSUE THREAD

**Issue by jethrogb** [openssl/openssl#21837]
OpenSSL's CMP code generates an incorrect AlgorithmIdentifier for PKIHeader.protectionAlg when using RSA. It doesn't emit any parameters, whereas it should emit a NULL parameter.

The problem is that the code that computes the `X509_ALGOR` for CMP always passes `V_ASN1_UNDEF` as the parameter:

https://github.com/openssl/openssl/blob/8c5284ff194f444877ae25012d3d07ee46e46219/crypto/cmp/cmp_protect.c#L228

However, as is clear from `ASN1_item_sign` you should be checking for `pkey->ameth->pkey_flags & ASN1_PKEY_SIGPARAM_NULL`:

https://github.com/openssl/openssl/blob/8c5284ff194f444877ae25012d3d07ee46e46219/crypto/asn1/a_sign.c#L246-L247

To produce a CMP message that exhibits the issue:

```
openssl genrsa 2048 > subjectkey.pem
openssl req -x509 -days 3650 -new -newkey rsa:2048 -keyout senderkey.pem -subj /CN=sender -out sendercert.pem -nodes
openssl cmp -cmd cr -reqout cr.der -newkey subjectkey.pem -subject "/CN=subject" -certout /dev/null -rspin /dev/null -recipient "/CN=recipient" -key senderkey.pem -cert sendercert.pem -extracerts sendercert.pem
openssl asn1parse -in cr.der -inform DER
```

```
    0:d=0  hl=4 l=1837 cons: SEQUENCE          
    4:d=1  hl=3 l= 146 cons:  SEQUENCE          
    7:d=2  hl=2 l=   1 prim:   INTEGER           :02
[...]
   74:d=2  hl=2 l=  13 cons:   cont [ 1 ]        
   76:d=3  hl=2 l=  11 cons:    SEQUENCE          
   78:d=4  hl=2 l=   9 prim:     OBJECT            :sha256WithRSAEncryption
[...]
  153:d=1  hl=4 l= 636 cons:  cont [ 2 ]        
  157:d=2  hl=4 l= 632 cons:   SEQUENCE          
  161:d=3  hl=4 l= 628 cons:    SEQUENCE          
  165:d=4  hl=4 l= 344 cons:     SEQUENCE          
[..]
  513:d=4  hl=4 l= 276 cons:     cont [ 1 ]        
  517:d=5  hl=2 l=  13 cons:      SEQUENCE          
  519:d=6  hl=2 l=   9 prim:       OBJECT            :sha256WithRSAEncryption
  530:d=6  hl=2 l=   0 prim:       NULL              
```
You can see here `PKIHeader.protectionAlg` is missing the NULL value, but it's there in `PKIBody.cr.0.popo.signature.algorithmIdentifier`.

**Comment by btme0011** [comment]
I am working on in.

**Comment by qunying** [comment]
Is this problem still actively being worked on?

**Comment by qunying** [comment]
The inclusion or not of this NULL parameter is discussed in RFC4055 section 2.1
```
All implementations MUST accept both NULL and absent parameters as
   legal and equivalent encodings.
```
And latter in section 5:
```
When any of these four object identifiers appears within an
   AlgorithmIdentifier, the parameters MUST be NULL.  Implementations
   MUST accept the parameters being absent as well as present.
```
In OSSL 3.2, I am kind of lost how to change the code, it used to work when  set_sig_algor() function still exist in older version:
```
--- crypto/cmp/cmp_protect.c
+++ crypto/cmp/cmp_protect.c
@@ -238,8 +238,19 @@ static int set_sig_algor(const OSSL_CMP_CTX *ctx, X509_ALGOR **alg)
     if (*alg == NULL && (*alg = X509_ALGOR_new()) == NULL)
         return 0;

-    if (X509_ALGOR_set0(*alg, algo, V_ASN1_UNDEF, NULL))
-        return 1;
+    switch (EVP_MD_nid(ctx->digest)) {
+    case NID_sha1:
+    case NID_sha224:
+    case NID_sha256:
+    case NID_sha384:
+    case NID_sha512:
+       // Add NULL parameter
+       if (X509_ALGOR_set0(*alg, algo, V_ASN1_NULL, NULL))
+           return 1;
+    default:
+       if (X509_ALGOR_set0(*alg, algo, V_ASN1_UNDEF, NULL))
+           return 1;
+    } // switch
     ASN1_OBJECT_free(algo);
     return 0;
 }
```

**Comment by rajeev-0** [comment]
As @qunying mentioned, RSA AlgorithmIdentifier with or without `NULL` parameter should be accepted. 
But any way, it is better to have uniformity and this got indirectly handled by commit #21884.

**Comment by t8m** [comment]
Closing then. Please note that this means it is not fixed on 3.0 and 3.1 but I think that is acceptable.
